src/main/java/org/jenkinsci/test/acceptance/po/PluginManager.java

```diff
@@ -213,35 +214,32 @@ public boolean installPlugins(final PluginSpec... specs) throws UnableToResolveD
                 }
             }
         } else {
-            visit("available");
+            List<PluginSpec> allRequiredPlugins = Arrays.asList(specs);
+            Map<Integer, List<String>> optionalDependencies = ucmd.get().optionalDependenciesOf(jenkins, allRequiredPlugins);
 
-            final ArrayList<PluginSpec> update = new ArrayList<>();
-            for (final PluginSpec n : specs) {
-                switch (installationStatus(n)) {
-                    case NOT_INSTALLED:
-                        tickPluginToInstall(n);
-                    break;
-                    case OUTDATED:
-                        update.add(n);
-                    break;
-                    case UP_TO_DATE:
-                        // Nothing to do
-                    break;
-                    default: assert false: "Unreachable";
+            for (List<String> optionalDependenciesLevel : optionalDependencies.values()) {
+                List<PluginSpec> optionalDependenciesToInstall = new ArrayList<>();
+                List<PluginSpec> specsWithoutOptional = new ArrayList<>();
+
+                for (PluginSpec requiredSpec : allRequiredPlugins) {
+                    if (optionalDependenciesLevel.contains(requiredSpec.getName())) {
+                        optionalDependenciesToInstall.add(requiredSpec);
+                    } else {
+                        specsWithoutOptional.add(requiredSpec);
+                    }
                 }
-            }
 
-            clickButton("Install");
+                allRequiredPlugins = specsWithoutOptional;
 
-            // Plugins that are already installed in older version will be updated
-            System.out.println("Plugins to be updated: " + update);
-            if (!update.isEmpty()) {
-                visit(""); // Updates tab
-                for (PluginSpec n : update) {
-                    tickPluginToInstall(n);
+                if (!optionalDependenciesToInstall.isEmpty()) {
+                    installOrUpdatePlugins(optionalDependenciesToInstall);
+
+                    // Jenkins will be restarted if necessary
+                    new UpdateCenter(jenkins).waitForInstallationToComplete(optionalDependenciesToInstall.toArray(new PluginSpec[]{}));
```

Comment:
If the final installation screen (that where a summary of all the installed plugins, including the optional ones and the other ones, appears) still displays the error that forces a restart, I think it would be fine restarting just then, as all the extension points should be functional afterwards.

If it doesn't, I'm afraid we need to restart here to be sure the plugin is available before the other one is installed.